### DISCUSSION TITLE
Warning SelectorRequired (selector is required)

### DISCUSSION THREAD

**Discussion by Moss-CG** [kedacore/keda#4342] [Q&A / Need Help]
I will preface this question with an apology for what I believe to be a simple error that I am just too unfamiliar with k8s to decipher.

 I have a custom resource that is handling setting up an opensearch cluster in my environment with pods for master/data/client nodes and I am trying to use KEDA to control scaling of the data nodes by CPU. Unfortunately when I apply the ScaledObject yaml, the ScaledObject looks fine, but the subsequent HPA that gets generated is unable to get the current metrics and has a target of <unknown> / targetPercentage.
![image](https://user-images.githubusercontent.com/1307487/223812760-72b9ae71-2455-4ed1-a5b2-ea1f3e4d6806.png)

When describing the HPA I see the following conditions and events:
![image](https://user-images.githubusercontent.com/1307487/223812361-73fd1f68-e6b8-4704-9d71-ac825a325832.png)

From my understanding I need a selector defined in my CRD that allows the HPA to get current metrics; but I am unsure where this needs to be added in my case. My custom resource yaml is formatted as follows:
![image](https://user-images.githubusercontent.com/1307487/223817103-77e03e5c-b743-4f34-9099-6182a3d67dca.png)

In the above image I tried adding a selector under nodes to no avail, as I am unsure how to tell the HPA to use that selector. 

Does anyone happen to have any suggestions for where I would need to add this selector to allow the HPA to get current metrics?

**Comment by tomkerkhove** [comment]
Hi, I'm not sure I get the scenario but CPU/memory are only supported on the scale target. If that scale target is a custom resource definition I'm not sure if we support it. 

Can you verify this @zroubalik?

**Reply by Moss-CG** [reply]
In this case the CRD exposes the scale subresource with the following targets:
![image](https://user-images.githubusercontent.com/1307487/224154982-2b0e4278-65cb-42ef-83b8-44c9f9834630.png)

The overall goal here is to be able to scale up/down Opensearch data nodes by CPU/Memory metrics. Possibly worth mentioning as well, I am able to scale by changing the minReplicaCount in the ScaledObject yaml, however when trying to trigger off CPU metrics my HPA has the issue above.

**Reply by tomkerkhove** [reply]
Oh I'm not questioning if CRD based scaling works, but more that it relies on CPU as it needs to get that somewhere.

**Reply by zroubalik** [reply]
This seems more like an issue in HPA 🤷‍♂️  

On the picture, I can see that in the `scale` you refernce the statusReplicasPath as `.status.nodes.replicas`, but the `status:` is empty. Can you double check?

**Reply by Moss-CG** [reply]
Yes status is empty in the subresources of my CRD but I also have it defined at the root level. I was attempting to follow [this scale subresource documentation](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource). Should I have the status subresource defined as something other than {}?
![image](https://user-images.githubusercontent.com/1307487/224723013-2a93b285-eefc-4922-874b-76d0dd23286f.png)

**Reply by zroubalik** [reply]
The json path needs to be valid and point to actual data in the CR not CRD. Also based on the error message it seems like you don't have the `/scale` subresource defined correct. See the `labelSelectorPath` item in the link you provided: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource


> It must be set to work with HPA.

**Reply by Moss-CG** [reply]
I originally tried passing `.spec.nodes.selector` to labelSelectorPath after adding a selector under nodes in the CR, however that then causes the ScaledObject check to fail stating that `ScaledObject doesn't have correct scaleTargetRef specification`; I thought since this was optional I could omit it or that I have defined my selector incorrectly. In the url example, `It must be set to work with HPA.`, is somewhat ambiguous. I am not sure what needs to be set to work with HPA in this case.

Inside the CR I have status defined as follows:
![image](https://user-images.githubusercontent.com/1307487/224761712-96e37883-53e5-425e-be5b-9cb66e28460c.png)

In this case I assumed the status subresource should be {} because it was at the root of the json.

**Reply by zroubalik** [reply]
The fields defined in `/scale` subresource should point to valid fields in spec & status.
